modules/crossover/src/lib.rs

```diff
@@ -21,66 +21,71 @@ pub struct Crossover {
-        let old_val = self.crossover;
-        uplink::debug!(
-            "setting crossover from {:?} to {:?}",
-            self.crossover,
-            to
-        );
-        self.crossover = to;
-        old_val
-    }
+        let self_is_initial = self.crossover == INITIAL_VALUE;
 
-    /// Test `set_crossover` functionality through a host transaction
-    pub fn self_call_test_a(self: &mut State<Self>, update: i32) -> i32 {
-        let old_value = self.crossover;
-        let callee = uplink::self_id();
-        self.transact::<_, i32>(callee, "set_crossover", &update)
-            .unwrap();
+        uplink::debug!("Querying module {module:?} for their state");
+        let other_crossover =
+            uplink::query::<_, i32>(module, "crossover", &value).unwrap();
+        let other_is_initial = other_crossover == INITIAL_VALUE;
 
-        assert_eq!(self.crossover, update);
-        old_value
-    }
+        self.set_crossover(value);
 
-    /// Test `set_crossover` functionality through a host raw transaction
-    pub fn self_call_test_b(
-        self: &mut State<Self>,
-        target: ModuleId,
-        raw_transaction: RawTransaction,
-    ) -> i32 {
-        let co = self.crossover;
-        self.set_crossover(co * 2);
-        self.transact_raw(target, &raw_transaction).unwrap();
-        self.crossover
+        self_is_initial && other_is_initial
     }
 
-    /// Update crossover and panic
-    pub fn update_and_panic(&mut self, new_value: i32) {
-        let old_value = self.crossover;
-        let callee = uplink::self_id();
+    // Sets the crossover to a new `value`, calls the calling contract - which
+    // is assumed to be another crossover contract - with "set_crossover" and
+    // panics afterwards.
+    pub fn set_call_and_panic(self: &mut State<Self>, value: i32) {
+        self.set_crossover(value);
 
-        // What should self.crossover be in this case?
+        let caller = uplink::caller();
+        uplink::debug!("calling back {caller:?}");
+        self.transact::<_, ()>(caller, "set_crossover", &value)
+            .unwrap();
 
-        // A: we live with inconsistencies and communicate them.
-        // B: we update self, which then should be passed to the transaction
+        uplink::debug!("panicking after setting the crossover");
+        panic!("OH NOES");
+    }
 
-        let q = uplink::query::<_, i32>(callee, "crossover", &new_value);
+    /// Return crossover value
+    pub fn crossover(&self) -> i32 {
+        uplink::debug!("Returning crossover: {}", self.crossover);
+        self.crossover
+    }
 
-        match q {
-            Ok(old) if old == old_value => panic!("OH NOES"),
-            _ => (),
-        }
+    /// Update crossover and return old value
+    pub fn set_crossover(&mut self, to: i32) -> i32 {
+        let old_val = self.crossover;
+        uplink::debug!("setting crossover from {old_val} to {to}");
+        self.crossover = to;
+        old_val
     }
 }
 
```

Comment:
As discussed, I'd use in this test contract different numbers so that it is easier and more convincing to follow state changes.

Reply:
I made it a little more clear by allowing the caller to pass all different values to (possibly) set.